### PR TITLE
Expand support for blueprint sources and add an action to import blueprints

### PR DESCRIPTION
**Pull request by luuquangvu**
## Summary by Sourcery

Add a blueprint import action with expanded source provider support and improved HTTP caching for blueprint updates.

New Features:
- Introduce an import_blueprint admin action and coordinator method to fetch, validate, and install blueprints from external URLs with safety checks.
- Add support for additional blueprint source providers including GitLab, Codeberg, Bitbucket, and a generic YAML URL handler with automatic metadata extraction.

Enhancements:
- Persist and propagate Last-Modified headers alongside ETags to optimize conditional HTTP requests and blueprint update logic.
- Expose provider type and related metadata on blueprint update entities and internal state for better introspection and compatibility.
- Align URL normalization, path generation, and error handling with Home Assistant Core blueprint importer semantics.

Documentation:
- Update English and Vietnamese README files with instructions and descriptions for importing blueprints and expanded source support.

Tests:
- Extend and add unit and integration tests for new import service flows, provider behavior, Last-Modified handling, and edge cases in the update coordinator.

## Summary by CodeRabbit

* **New Features**
  * Import blueprints by URL from many providers with explicit confirmation, automatic metadata extraction, stable install paths, provider type reporting, and ETag/Last-Modified support.

* **Documentation**
  * Added "Importing Blueprints" guide; clarified supported sources, import workflow, restore behavior, and sandboxed impact details.

* **Translations**
  * UI labels and import/error messages added/updated across multiple languages.

* **Tests**
  * New and expanded tests for import service, providers, last-modified, storage/restore, template support, and edge cases.

* **Bug Fixes**
  * Improved fetch/update robustness, validation, and update detection.

[Review Change Stack](https://app.coderabbit.ai/change-stack/luuquangvu/blueprints-updater/pull/75)

### CODE REVIEW

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Review by luuquangvu — COMMENTED** [review]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@sourcery-ai review

File: custom_components/blueprints_updater/coordinator.py [hunk 1]
```diff
@@ -1070,6 +1077,133 @@ async def async_reload_services(self, domains: list[str] | set[str] | None = Non
...
+        if full_path in self.data:
+            existing_url = self.data[full_path].get("source_url")
+        elif os.path.exists(full_path):
+            try:
+                with open(full_path, encoding="utf-8") as f:
+                    content_on_disk = f.read()
+                parsed_disk = yaml_util.parse_yaml(content_on_disk)
+                if isinstance(parsed_disk, dict):
+                    existing_url = parsed_disk.get("blueprint", {}).get("source_url")
+            except Exception:
```
Comment (SourceryAI) [inline custom_components/blueprints_updater/coordinator.py]: **suggestion (bug_risk):** Avoid bare `except Exception` when reading existing blueprint file

Swallowing all exceptions here (including permission/IO, decoding, or YAML errors) makes configuration problems invisible and can hide real issues with the existing file. Please catch only the expected exceptions (e.g., `OSError`, `UnicodeDecodeError`, YAML/HomeAssistant-specific errors) and at least log a debug message with the file path so failures to determine `existing_url` are traceable.

Suggested implementation:

```python
        rel_path = f"{domain}/{author}/{name}.yaml"
        full_path = self.hass.config.path(BLUEPRINTS_DATA_DIR, rel_path)

        existing_url = None
        if full_path in self.data:
            existing_url = self.data[full_path].get("source_url")
        elif os.path.exists(full_path):
            try:
                with open(full_path, encoding="utf-8") as f:
                    content_on_disk = f.read()
                parsed_disk = yaml_util.parse_yaml(content_on_disk)
            except (OSError, UnicodeDecodeError, yaml_util.YamlException) as err:
                _LOGGER.debug(
                    "Failed to read existing blueprint file %s to determine source_url: %s",
                    full_path,
                    err,
                )
            else:
                if isinstance(parsed_disk, dict):
                    existing_url = parsed_disk.get("blueprint", {}).get("source_url")

```

1. Ensure this module has a logger defined, e.g. near the top of the file:
   `import logging` and `_LOGGER = logging.getLogger(__name__)` if not already present.
2. Ensure `yaml_util` is imported from Home Assistant’s YAML utility module and supports `YamlException`, typically:
   `from homeassistant.util import yaml as yaml_util`.
3. Confirm `os` is imported (`import os`) at the top of the file if it is not already.

✅ Addressed in 696cacc277404db2d81eba2dcb3d15da35c61d8d: The bare except around reading the existing blueprint file has been replaced with a narrow catch of (OSError, UnicodeDecodeError, HomeAssistantError) and now logs a debug message with the file path when parsing fails.

**Review by SourceryAI — COMMENTED** [review]
Hey - I've found 1 issue, and left some high level feedback:

- In `async_import_blueprint`, the safety check for generic URLs relies solely on the `Content-Type` header with `text/plain` allowed, which could let non‑YAML payloads (e.g., HTML served as text/plain) through; consider tightening this by attempting a YAML parse first and/or adding a simple heuristic (e.g., requiring a top‑level `blueprint:` key) before proceeding.
- The import logic checks `self.data` using `full_path` as the key; if other parts of the coordinator consistently key `self.data` by relative path (or vice versa), this lookup will never hit and could silently bypass the existing‑URL conflict detection—worth double‑checking that path conventions are aligned here.

<details>
<summary>Prompt for AI Agents</summary>

~~~markdown
Please address the comments from this code review:

## Overall Comments
- In `async_import_blueprint`, the safety check for generic URLs relies solely on the `Content-Type` header with `text/plain` allowed, which could let non‑YAML payloads (e.g., HTML served as text/plain) through; consider tightening this by attempting a YAML parse first and/or adding a simple heuristic (e.g., requiring a top‑level `blueprint:` key) before proceeding.
- The import logic checks `self.data` using `full_path` as the key; if other parts of the coordinator consistently key `self.data` by relative path (or vice versa), this lookup will never hit and could silently bypass the existing‑URL conflict detection—worth double‑checking that path conventions are aligned here.

## Individual Comments

### Comment 1
<location path="custom_components/blueprints_updater/coordinator.py" line_range="1173-1180" />
<code_context>
+        existing_url = None
+        if full_path in self.data:
+            existing_url = self.data[full_path].get("source_url")
+        elif os.path.exists(full_path):
+            try:
+                with open(full_path, encoding="utf-8") as f:
+                    content_on_disk = f.read()
+                parsed_disk = yaml_util.parse_yaml(content_on_disk)
+                if isinstance(parsed_disk, dict):
+                    existing_url = parsed_disk.get("blueprint", {}).get("source_url")
+            except Exception:
+                pass
+
</code_context>
<issue_to_address>
**suggestion (bug_risk):** Avoid bare `except Exception` when reading existing blueprint file

Swallowing all exceptions here (including permission/IO, decoding, or YAML errors) makes configuration problems invisible and can hide real issues with the existing file. Please catch only the expected exceptions (e.g., `OSError`, `UnicodeDecodeError`, YAML/HomeAssistant-specific errors) and at least log a debug message with the file path so failures to determine `existing_url` are traceable.

Suggested implementation:

```python
        rel_path = f"{domain}/{author}/{name}.yaml"
        full_path = self.hass.config.path(BLUEPRINTS_DATA_DIR, rel_path)

        existing_url = None
        if full_path in self.data:
            existing_url = self.data[full_path].get("source_url")
        elif os.path.exists(full_path):
            try:
                with open(full_path, encoding="utf-8") as f:
                    content_on_disk = f.read()
                parsed_disk = yaml_util.parse_yaml(content_on_disk)
            except (OSError, UnicodeDecodeError, yaml_util.YamlException) as err:
                _LOGGER.debug(
                    "Failed to read existing blueprint file %s to determine source_url: %s",
                    full_path,
                    err,
                )
            else:
                if isinstance(parsed_disk, dict):
                    existing_url = parsed_disk.get("blueprint", {}).get("source_url")

```

1. Ensure this module has a logger defined, e.g. near the top of the file:
   `import logging` and `_LOGGER = logging.getLogger(__name__)` if not already present.
2. Ensure `yaml_util` is imported from Home Assistant’s YAML utility module and supports `YamlException`, typically:
   `from homeassistant.util import yaml as yaml_util`.
3. Confirm `os` is imported (`import os`) at the top of the file if it is not already.
</issue_to_address>
~~~

</details>

***

Hi @luuquangvu! 👋

Thanks for trying out Sourcery by commenting with `@sourcery-ai review`! 🚀

Install the [sourcery-ai bot](https://github.com/apps/sourcery-ai/installations/new) to get automatic code reviews on every pull request ✨

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by SourceryAI** [comment]
Sorry @luuquangvu, you have reached your weekly rate limit of 500000 diff characters.

Please try again later or [upgrade](https://app.sourcery.ai/login?connection=github&git_namespace_id=6678374&utm_source=github&utm_medium=bot&utm_campaign=rate_limited) to continue using Sourcery

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by SourceryAI** [comment]
Sorry @luuquangvu, you have reached your weekly rate limit of 500000 diff characters.

Please try again later or [upgrade](https://app.sourcery.ai/login?connection=github&git_namespace_id=6678374&utm_source=github&utm_medium=bot&utm_campaign=rate_limited) to continue using Sourcery

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai full review

File: custom_components/blueprints_updater/providers.py [hunk 2]
```diff
@@ -224,6 +335,136 @@ def parse_content(
...
+        return _replace_path_segment(url, "/-/raw/", "/-/blob/", "/-/raw/")
+
+    def get_metadata(self, url: str, content: str | None = None) -> dict[str, str]:
+        """Extract metadata from GitLab URL (Matching HA Generic Logic)."""
+        parsed = urlparse(url)
+        author = parsed.hostname.lower() if parsed.hostname else "imported"
+        path_parts = parsed.path.strip("/").split("/")
+        filename = path_parts[-1] if path_parts else "blueprint.yaml"
+        name = filename.removesuffix(".yaml").removesuffix(".yml")
+        return {"author": author, "name": name}
```
[the remark this answers is not in the document]
Comment (luuquangvu) [reply custom_components/blueprints_updater/providers.py:376]: This project requires 100% parity with Home Assistant Core's internal importer. Including repo in the path would cause blueprints to be stored in different locations than the standard HA UI, leading to duplicate entries and broken update management. We prefer handling collisions via explicit conflict errors rather than deviating from the standard HA directory structure.

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@coderabbitai full review
